pkg/sqlcmd/sqlcmd.go

```diff
@@ -284,7 +287,12 @@ func (s *Sqlcmd) IncludeFile(path string, processAll bool) error {
 	}
 	defer f.Close()
 	b := s.batch.batchline
-	scanner := bufio.NewScanner(f)
+	win16be := unicode.UTF16(unicode.BigEndian, unicode.IgnoreBOM)
+	utf16bom := unicode.BOMOverride(win16be.NewDecoder())
+	unicodeReader := transform.NewReader(f, utf16bom)
+	scanner := bufio.NewScanner(unicodeReader)
+	buf := make([]byte, maxLineBuffer)
+	scanner.Buffer(buf, maxLineBuffer)

```

Comment:
I tried making this change locally and it breaks reading utf8 input.

Reply:
aha... thx for info. I'm now looking the way how to running tests on my M1. Could be that i will try once more time on this weekend.

Reply:
![image](https://user-images.githubusercontent.com/2224906/166304975-100e7ec6-7105-4383-9eee-2787c60997b1.png)

Shows the LE BOM, and in the debugger I see `bomOverride` read it correctly and set `d.current` to a `utf16le.NewDecoder()`

```go
func (d *bomOverride) Transform(dst, src []byte, atEOF bool) (nDst, nSrc int, err error) {
	if d.current != nil {
		return d.current.Transform(dst, src, atEOF)
	}
	if len(src) < 3 && !atEOF {
		return 0, 0, transform.ErrShortSrc
	}
	d.current = d.fallback
	bomSize := 0
	if len(src) >= 2 {
		if src[0] == 0xFF && src[1] == 0xFE {
====>		d.current = utf16le.NewDecoder()
			bomSize = 2
		} else if src[0] == 0xFE && src[1] == 0xFF {
   	d.current = utf16be.NewDecoder()
			bomSize = 2
		} else if len(src) >= 3 &&
			src[0] == utf8BOM[0] &&
			src[1] == utf8BOM[1] &&
			src[2] == utf8BOM[2] {
			d.current = transform.Nop
			bomSize = 3
		}
	}
	if bomSize < len(src) {
		nDst, nSrc, err = d.current.Transform(dst, src[bomSize:], atEOF)
	}
	return nDst, nSrc + bomSize, err
}
```

Reply:
ah ha my input file was missing a `N` around the string constant.

Reply:
In summary - if it's ok to require a utf16 BOM, we can get away with my change to just use BOMOverride and fallback to the UTF8 decoder.
Otherwise, we would need an additional scan of the file first, or at least a partial scan, to try to detect if it's BOM-less UTF16 file.

Reply:
Looking at the ODBC sqlcmd code, I think we can assume UTF16 files have a BOM. Unlike the ODBC one, go-sqlcmd will not use any Windows code page translations as the fallback. 

This is all ODBC sqlcmd does when opening the file:

```C++
HRESULT CTextStream::ReadHeaderEncoding()
{
    BYTE buffer[4] = {0};
    DWORD bytesRead = 0 ;
    m_StartOffset = 0;

    m_FileSize = ::GetFileSize(m_FileHandle, NULL);
    if (m_FileSize != INVALID_FILE_SIZE && m_FileSize >= 4)
    {
        //check for BE,LE,UTF8
        if (!ReadFile(m_FileHandle, buffer, 4, &bytesRead, NULL))
            return HRESULT_FROM_WIN32(GetLastError());

        //little endian unicode file
        if (bytesRead >= 2 && memcmp(buffer, LEStart, 2) == 0)
        {
            //size of header
            m_StartOffset = 2;
            m_CodePage = -1;
        }
        //big endian unicode file
        else if (bytesRead >= 2 && memcmp(buffer, BEStart, 2) == 0)
        {
            //size of header
            m_StartOffset = 2;
            m_bLittleEndian = false;
            m_CodePage = -1;
        }
        //UTF8 format file
        else if (bytesRead >= 3 && memcmp(buffer, UTF8Start, 3) == 0)
        {
            //size of header
            m_StartOffset = 3;
            m_CodePage = CP_UTF8;
        }

        ::SetFilePointer(m_FileHandle, m_StartOffset, NULL, FILE_BEGIN);
    }

    InitFileCache();

    return S_OK;
}
```